src/Database/Franz/Server.hs

```diff
@@ -0,0 +1,186 @@
+{-# LANGUAGE DeriveGeneric, LambdaCase, OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE BangPatterns #-}
+module Database.Franz.Server
+  ( Settings(..)
+  , startServer
+  , defaultPort
+  ) where
+
+import Control.Concurrent
+import Control.Exception
+import Control.Monad
+import Control.Retry
+import Control.Concurrent.STM
+import Database.Franz.Internal
+import Database.Franz.Protocol
+import Database.Franz.Reader
+import Data.ConcurrentResourceMap
+import Data.Serialize
+import qualified Data.IntMap.Strict as IM
+import Data.IORef
+import qualified Data.ByteString.Char8 as B
+import qualified Data.HashMap.Strict as HM
+import Data.Tuple (swap)
+import qualified Network.Socket.SendFile.Handle as SF
+import qualified Network.Socket.ByteString as SB
+import qualified Network.Socket as S
+import System.Directory
+import System.FilePath
+import System.IO
+import System.Process (ProcessHandle, spawnProcess, cleanupProcess,
+  waitForProcess, getProcessExitCode, getPid)
+
+respond :: FranzReader
+  -> IORef (IM.IntMap ThreadId)
+  -> FilePath
+  -> IORef B.ByteString
+  -> MVar S.Socket -> IO ()
+respond env refThreads path buf vConn = do
+  recvConn <- readMVar vConn
+  runGetRecv buf recvConn get >>= \case
+    Right (RawRequest reqId req) -> do
+      (stream, query) <- handleQuery env path req
+      touchActivity stream
```

Comment:
This approach changes the broken user-counting behaviour in favour of the stream possibly being randomly closed while it's still being used. If reaper runs on a unused stream before we manage to touch it here, it's going to close it and we end up with broken reference to stream. A tempting change may be to `touchActivity` while having the lock inside `handleQuery`: this reduces the likelyhood of dead handle but it's still technically a race condition; if handling query takes longer than reaper (which user can set to low interval), we end up with same issue.

We still want to increase ref count (really, we don't care about the number, just whether it's in use or not) before giving up lock on `vStream` in `handleQuery` to `stream` and we just need to ensure that this count is decreased at stream use-site.

The code is complicated by having a case where we we fork conditionally. Could we handle all the requests in a thread? This would let us _uniformly_ handle the stream resource. Something like

```
forkIO $ bracket handleQuery (\(stream, _) -> removeActivity stream) $ \(stream, query) -> do
  {- handle request as necessary -}
```

This would make the system more responsive with multiple requests too, no?
It doesn't have to be this way, we could restructure the code to make sure in each branch we clean up the resource properly but it gets messy I think...

Reply:
This only happens when the handler is running more than`streamLifetime` without activity counter right? That's very unlikely

Reply:
>This only happens when the handler is running more than `streamLifetime` without activity counter right? That's very unlikely

Not for the current code: say, `streamLifetime` passes since we last used stream but reaper hasn't ran just yet. We enter `handleQuery` and return the stream. Before we hit `touchActivity` in `respond`, reaper runs and closes the stream.

If you move `touchActivity` into the lock in `handleQuery` then you end up with race condition that instant query has to finish under `streamLifetime` and busy-queries have to start handling (through the thread) under `streamLifetime`.

I don't think it's very unlikely at all: these things take locks, do IO, and user can trivially set `streamLifetime` to `0` because they decide they don't want to let resources to sit around. Then you're at complete mercy of `reapInterval`.

It's kind of the definition of race condition...

Reply:
Forking a thread for each request made the performance worse. Still, it was straightforward to ensure activity counter increments by making `handleQuery` bracket style.

Reply:
Right, this is a lot better (this is exactly the concurrent resource map approach).